### ISSUE TITLE
Proposal for dependency tracking with 'fstardep'

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#438]
With several people, including @msprotz , @fournet, @mlr-msft, @prosecco  and @parno, I've had conversations about a more sensible and automatic way of computing dependencies of an F\* module. I'm describing the proposal here. The time has come to actually implement this. Comments are most welcome, of course. 
1. A 1-1 correspondence between F\* modules/interfaces and filenames

The implementation of module M will be in file M.fst. M may itself be a long name. For example, we may have FStar.Heap.fst. 

The interface of a module M is optional. If present, it will be in a file called M.fsti.

Of course, this means 1 module per file.
1. A tool called 'fstardep' which can be run as

`fstardep -I path1 -I path2 ... -I pathn M1.fst ... Mn.fst`

It parses `M1.fst` and resolves each name in module `M1` to the module or interface `N` which provides that name. Module `M1` depends on each such `N`. This is repeated for each module up to `Mn`.

`fstardep` resolves such dependences by locating `N.fst(i)` in exactly one of the include-paths `[path1; ...; pathn]`, in that order.  

For each such dependence, `fstardep` transitively computes its dependences. 

Its output is a dependency graph emitted in a Makefile-compatible syntax. For example, it may say:

`M.fst: prims.fst N.fst N.fsti`

---

I plan to implement this soon. The implementation work should be relatively easy. What will be more painful is migrating our samples and libraries to comply with the naming restrictions mentioned above. Initially, let's not enforce this naming restriction---for files that comply with it, they benefit from the fstardep functionality. Eventually, let's move all our files to this convention and from then on enforce the convention strictly.

**Comment by nikswamy** [comment]
BTW, this also means that `build-config` will go away.

**Comment by aseemr** [comment]
So what is the usage model with this change. For batch mode verification, first run fstardep and then use the generated graph to build the fstar command. But what about interactive verification ?

**Comment by aseemr** [comment]
Also it would be nice to have some default directories in the -l path (such as the current directory and lib/ directory).

**Comment by nikswamy** [comment]
For interactive verification, many of us already use this bit of magic to get the emacs mode to work well with a makefile.
https://github.com/mitls/mitls-fstar#configuring-emacs-and-atom-f-modes

**Comment by nikswamy** [comment]
Including some default directories in the include path is a good idea. Current dir and lib are good suggestions; will do.

**Comment by mlr-msft** [comment]
- Unless `fstar.exe` actually produces an output named `M.fst.ver`, I would recommend to not include it in the `fstardep` output because it seems to me like a build-system specific target. For example, `NuBuild` creates verification reports from the output of the tool and has its own method for creating and naming targets for these reports automatically. `fstar.exe` wanting to be the source of these target names seems to complicate things.
- `Makefile` syntax is convenient for use cases involving `gmake` but it seems to me that it would make other use cases more burdensome, such as interactive verification where you might prefer to scan and feed the dependency list right back into `fstar.exe`. `NuBuild` and `rake` are other situations where  `Makefile` syntax might not be the best choice. A simple `\n` separated list of files works for single module targets, which suits all of `NuBuild`'s needs. A common structured format with readily available parsers (such as JSON) seems more appropriate to me if we want to be able describe multiple targets and their associated dependencies.

**Comment by s-zanella** [comment]
I'd prefer the default output to be a simple list of space separated file names. It's easy to add command-line options for additional output formats, including one that is suitable to use with `gmake`.

It would be nice if `fstardep` could process multiple source files in a single run, much like `ocamldep`, so that this simple Makefile pattern is supported:

```
FSTARDEP=fstardep --gmake

depend:
        $(FSTARDEP) $(INCLUDE) *.fsti *.fst > .depend

include .depend
```

**Comment by mlr-msft** [comment]
> I'd prefer the default output to be a simple list of space separated file names.

I would agree, except file names may contain spaces themselves. Giving spaces a dual-role in the syntax leads to the need for meta-characters and escaping. This output is likely to be hand parsed and any nuance in the syntax would likely go addressed, leading to bugs.

I think that newline separators preserve the simplicity of your suggestion without the hidden costs of using spaces.

>  It's easy to add command-line options for additional output formats, including one that is suitable to use with  gmake .

I hadn't considered adding support for multiple output formats. I like that. ^-^

**Comment by nikswamy** [comment]
No spaces in file names. File names correspond to module identifiers, and the space character is not a valid identifier symbol.

**Comment by mlr-msft** [comment]
Oops! I should have used the word _path_ instead of _file_.

e.g. `C:\My Modules\FStar.Seq.fst`

**Comment by nikswamy** [comment]
(Sigh.)
Ok, you're right. 
What does ocamldep do about spaces in paths? Escapes them, presumably.

**Comment by mlr-msft** [comment]
The example provided [here](http://caml.inria.fr/pub/docs/manual-ocaml/depend.html) suggests that `ocamldep` works like @s-zanella's `fstardep --gmake` proposal. So, I assume that it escapes spaces in the manner prescribed by `gmake` syntax.

I don't have experience with `ocamldep`, however, so hopefully someone else can elaborate.

**Comment by s-zanella** [comment]
Indeed, `ocamldep` escapes spaces in paths:
```
$ ocamldep -I "../My Modules/" M.mli
M.cmi : ../My\ Modules/N.cmi O.cmi P.cmi
```
But I just learned that escaping spaces in a GNU make is a long-standing problem (http://savannah.gnu.org/bugs/?712), so I don't know how well will this work on different platforms and on BSD make / GNU make.

**Comment by mlr-msft** [comment]
Yes, it is frustrating. I experimented a long time ago with developing my own workarounds for this issue in BSD & GNU make and failed to find something really satisfying. :(

If we provide the newline-separated format being discussed, would it then be sufficient for the output of `fstar.exe --gmake` to simply mimic `ocamldep`?

**Comment by s-zanella** [comment]
That would be sufficient for our current use of make in miTLS.

**Comment by nikswamy** [comment]
Updated the original post with some points from Jonathan.

**Comment by msprotz** [comment]
Some notes to self:

- `open` directives always refer to an absolute path, not a relative one.
For instance, the following does not open `FStar.Heap` in the scope.
```
open FStar
open FStar.heap
```

- The tool must output the transitive closure of the dependency graph because there is no support for separate verification.
- One can either adopt a style in which one invocation of fstar verifies everything at once, with just one target in the Makefile (the dependency graph is not encoded in the makefile); or, one can adopt a style in which all intermediary targets appear in the makefile, and only `target.fs` is fully verified (all other dependencies are "lightly verified")
- Generating that Makefile would be great
- We want an --auto mode, where everything is discovered automatically by F*.
  - The variant where everything is fully verified would be used from the command-line, for simple projects without a makefile.
  - The variant where dependencies are lightly verified would be used for the interactive mode.
- Once the --auto mode is implemented, we'll be able to get rid of build-config.

**Comment by msprotz** [comment]
Status update. `fstar` now takes a `--dep X` option where `X` is either `make` or `nubuild`.

In the former case, it prints in a Makefile format the dependencies of all files passed on the command-line. Targets are printed as they were passed to fstar. Dependencies are always basenames (no paths). It assumes that any -I option given to `fstar --dep` will also be given to subsequent invocations of fstar.

In the latter case, it prints the dependencies of the last argument passed on the command-line, separated by newlines.

- Because `dsenv` prepends a lot of filenames, in the case of the `make` format, you will see several targets, such as `FStar.{All,Set,Heap,ST}.fst` -- this is undesirable.
- The goal is to have an `--auto-deps` option that runs `fstar --dep`, then processes any dependencies; once this is done, we will be able to remove the code that prepends extra filenames to the command-line. This option will be enabled by default so that "fstar foo.fst" works out of the box without any voodoo.
- Once this option is implemented, we can make build-config go away.
- The is currently a conflict between the field projection syntax (`Some.v`) and the module prefixing syntax (`MyModule.f`); one thought Nik and I had was to use remove `Some.v` and use `e.v`, possibly enabling `e.Some.v` in case multiple branches of the data-type `e` belongs to have parameters that shadow each other. That doesn't work because it clashes with the record field syntax (`e.Othermodule.blah`). We don't want to type-check the file to perform dependency analysis.
- You will get warnings if you file does not follow the naming convention. Please fix it if you have the courage!
